### ISSUE TITLE
How to increase model performance?

### ISSUE THREAD

**Issue by EricLBuehler** [huggingface/candle#1834]
Hello all,

I have recently benchmarked completion token time, which is 30ms on an H100. However, with llama.cpp it is 10ms. Because [mistral.rs](https://github.com/EricLBuehler/mistral.rs) is built on Candle, it inherits this performance deficit. In #1680, @guoqingbao said that the Candle implementation is not suitable for batched computing because of naive CUDA kernels. What other areas could be optimized?

**Comment by LaurentMazare** [comment]
Could you provide a bit more details, e.g. what the benchmark actually do, profiling information so that we see where the time is actually spent?
I think a good example of a library achieving good performance results with candle is [text-embeddings-inference](https://github.com/huggingface/text-embeddings-inference) so you may want to have a look at the techniques/custom kernels that they use.

**Comment by guoqingbao** [comment]
> Could you provide a bit more details, e.g. what the benchmark actually do, profiling information so that we see where the time is actually spent? I think a good example of a library achieving good performance results with candle is [text-embeddings-inference](https://github.com/huggingface/text-embeddings-inference) so you may want to have a look at the techniques/custom kernels that they use.

From my experience for porting candle to a custom DSA, I found the bottle neck of candle is the implementation of CUDA kernels. The copy_strided_src use ucopy to process transpose, narrow, broadcast, etc. which is not efficient because it need to compute every input indices for the output. I have modified Layout to trace the changes of tensor layout and there is a transform type added which will send to the kernel and the ucopy kernel will do direct d2d transpose, broadcast, narrow (slice) without using indice compute and the existing approach (indice compute) is used as the last resort. Besides, I have replaced the build-in ops (frequently used ops in LLM such as layernorm, softmax, rotary embedding, etc.) with fused kernels (same interface but launching a custom kernel) and the performance was doubled. Another notable change is the kv concat (which is pretty low performance when generating long sentences) and we replaced it with custom kvconcat kernel, the kvconcat performs d2d copy instead of ucopy (compute indices) and the resulting effect is the generation was not being slowed down significantly in the long sentence generation. Other modification are removing unnecessary memsets, using inplace ops for unary ops and L2 cache for matmul weights.
